### PR TITLE
Mosh RAI changes

### PR DESCRIPTION
**Pull request by trantanen**
- Support for RAI for ping
- Move global RAI enable from 'sock' to 'link'

### CODE REVIEW

**Review by tokangas — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 1]
```diff
@@ -532,3 +532,24 @@ int link_func_mode_set(enum lte_lc_func_mode fun)
 
 	return return_value;
 }
+
+int link_rai_enable(bool enable)
+{
+	enum at_cmd_state state = AT_CMD_ERROR;
+	char command[] = "AT%%RAI=0";
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link.c:539]: Changed to reserve 10 bytes. Seems flash is reduced by 24 bytes when the change is made and doesn't matter if reservation is 10 or 9 bytes.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 2]
```diff
@@ -105,7 +106,9 @@ static const char link_usage_str[] =
 	"                Has impact after the bootup\n"
 	"  edrx:         Enable/disable eDRX with default or with custom parameters\n"
 	"  psm:          Enable/disable Power Saving Mode (PSM) with\n"
-	"                default or with custom parameters\n";
+	"                default or with custom parameters\n"
+	"  rai:          Enable/disable RAI feature. Actual use must be set for sockets\n"
+	"                with 'sock rai' command.";
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:111]: It's mentioned in the message when you enable/disable RAI.

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 3]
```diff
@@ -532,3 +532,24 @@ int link_func_mode_set(enum lte_lc_func_mode fun)
...
+	err = at_cmd_write(command, NULL, 0, &state);
+	if (state == AT_CMD_OK) {
+		shell_print(
+			shell_global,
+			"Release Assistance Indication functionality set to enabled=%d.\n"
+			"The change will be applied when going to normal mode for the next time.",
+			enable);
+	} else {
+		shell_error(shell_global, "Error state=%d, error=%d", state, err);
+		return -EINVAL;
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link.c:552]: True. Changed to EFAULT.

**Review by jhirsi — COMMENTED** [review]
Lookks good, couple issues still.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 4]
```diff
@@ -260,6 +263,12 @@ static const char link_tau_usage_str[] =
 	"                         Default value from\n"
 	"                         CONFIG_LTE_LC_TAU_PRE_WARNING_TIME_MS.\n";
 
+static const char link_rai_usage_str[] =
+	"Usage: link rai --enable | --disable\n"
+	"Options:\n"
+	"  -d, --disable,      Disable RAI\n"
+	"  -e, --enable,       Enable RAI\n";
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:271]: added -r, --read

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 5]
```diff
@@ -1283,6 +1298,13 @@ int link_shell(const struct shell *shell, size_t argc, char **argv)
 	case LINK_CMD_CONNECT:
 		ret = link_shell_pdn_connect(shell, apn, family);
 		break;
+	case LINK_CMD_RAI:
+		if (link_cmd_args.common_option == LINK_COMMON_ENABLE) {
+			link_rai_enable(true);
+		} else {
+			link_rai_enable(false);
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:1308]: LINK_COMMON_DISABLE check added

File: samples/nrf9160/modem_shell/src/ping/icmp_ping_shell.c [hunk 6]
```diff
@@ -29,6 +29,7 @@ static const char icmp_ping_shell_cmd_usage_str[] =
 	"  -I, --cid, [int]         Use this option to bind pinging to specific CID.\n"
 	"                           See link cmd for interfaces\n"
 	"  -6, --ipv6,              Force IPv6 usage with the dual stack interfaces\n"
+	"  -r, --rai                Set RAI options for ping socket\n"
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/ping/icmp_ping_shell.c]: Added comment about link rai to the help. Also added warning if rai is disabled when ping or socket rai options are used.

File: samples/nrf9160/modem_shell/src/link/link_api.c [hunk 7]
```diff
@@ -890,3 +890,57 @@ void link_api_modem_info_get_for_shell(const struct shell *shell, bool connected
...
+	}
+	*rai_status = false;
+
+	err = at_cmd_write(AT_CMD_RAI, at_response_str, sizeof(at_response_str), &state);
+	if (state != AT_CMD_OK) {
+		shell_error(shell_global, "Error state=%d, error=%d", state, err);
+		return -EFAULT;
+	}
+
+	err = at_params_list_init(&param_list, AT_CMD_RAI_RESP_PARAM_COUNT);
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link_api.c:917]: The list allocated by `at_params_list_init()` is not freed using `at_params_list_free()`, so there's memory leak.

**Review by tokangas — CHANGES_REQUESTED** [review]

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 8]
```diff
@@ -625,6 +625,24 @@ int link_func_mode_set(enum lte_lc_func_mode fun)
...
+	if (err == 0) {
+		shell_print(
+			shell_global,
+			"Release Assistance Indication status is enabled=%d",
+			rai_status);
+	} else {
+		shell_error(shell_global, "Reading RAI failed with error code %d", err);
+	}
+
+	return 0;
```
[the remark this answers is not in the document]
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link.c:643]: void is fine here. Return was a copy-paste leftover.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 9]
```diff
@@ -100,7 +101,9 @@ static const char link_usage_str[] =
 	"                Has impact after the bootup\n"
 	"  edrx:         Enable/disable eDRX with default or with custom parameters\n"
 	"  psm:          Enable/disable Power Saving Mode (PSM) with\n"
-	"                default or with custom parameters\n";
+	"                default or with custom parameters\n"
+	"  rai:          Enable/disable RAI feature. Actual use must be set for sockets\n"
+	"                with 'sock rai' command. Effective when going to normal mode.\n";
```
Comment (jhirsi) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:106]: If you want to say about the actual usage, then ping should be also mentioned in addition to mentioned "sock rai". IMO, you could just say more generically about the actual usage.

Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:106]: I'll write generally:
Actual use must be set for commands supporting RAI

File: samples/nrf9160/modem_shell/src/link/link_api.c (hunk 7 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_api.c:917]: Aargh, fixed.

**Review by tokangas — APPROVED** [review]

**Review by jhirsi — APPROVED** [review]

**Comment by trantanen** [comment]
@rlubos This is ready from our side and pending your review.

**Review by rlubos — APPROVED** [review]

**Pull request merged**